internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1120 @@ package controller
+		r.log.Error(err, "Failed to query the existence of a CSI Driver")
+		return err
+	}
+	if csiDriver.UID != "" {
+		ownerObjKey := client.ObjectKey{}
+		if ownerRef := csiDriver.GetAnnotations()[ownerRefAnnotationKey]; ownerRef != "" {
+			if err := json.Unmarshal([]byte(ownerRef), &ownerObjKey); err != nil {
+				r.log.Error(err, "Failed to parse owner ref annotation on CSI Driver")
+				return err
+			}
+		}
+		if csiDriver.Namespace != r.driver.Namespace || csiDriver.Name != r.driver.Name {
+			err := fmt.Errorf("invalid driver name")
+			r.log.Error(err, "Desired name already in use by a different CSI Driver")
+		}
+	}
+
+	// Load operator configuration resource
+	opConfig := csiv1a1.OperatorConfig{}
+	opConfig.Name = operatorConfigName
+	opConfig.Namespace = operatorNamespace
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&opConfig), &opConfig); client.IgnoreNotFound(err) != nil {
+		r.log.Error(err, "Unable to load operatorconfig.csi.ceph.io", "name", client.ObjectKeyFromObject(&opConfig))
+		return err
+	}
+
+	// Extract the driver sort name and driver type
+	matches := nameRegExp.FindStringSubmatch(r.driver.Name)
+	if len(matches) != 3 {
+		return fmt.Errorf("invalid driver name")
+	}
+	r.driverPrefix = matches[1]
+	r.driverType = DriverType(strings.ToLower(matches[2]))
+
+	// Load the current desired state in the form of a ceph csi driver resource
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&r.driver), &r.driver); err != nil {
+		r.log.Error(err, "Unable to load driver.csi.ceph.io object", "name", client.ObjectKeyFromObject(&r.driver))
+		return client.IgnoreNotFound(err)
+	}
+
+	// Creating a copy of the driver spec, making sure any local changes will not effect the object residing
+	// in the client's cache
+	r.driver.Spec = *r.driver.Spec.DeepCopy()
+	mergeDriverSpecs(&r.driver.Spec, opConfig.Spec.DriverSpecDefaults)
+
+	r.images = maps.Clone(imageDefaults)
+	if r.driver.Spec.ImageSet != nil {
+		imageSetCM := corev1.ConfigMap{}
+		imageSetCM.Name = r.driver.Spec.ImageSet.Name
+		imageSetCM.Namespace = r.driver.Namespace
+		if err := r.Get(r.ctx, client.ObjectKeyFromObject(&imageSetCM), &imageSetCM); err != nil {
+			r.log.Error(err, "Unable to load driver specified image set config map", "name", client.ObjectKeyFromObject(&imageSetCM))
+			return err
+		}
+
+		maps.Copy(r.images, imageSetCM.Data)
+	}
+
+	return nil
+}
+
+func (r *driverReconcile) reconcileK8sCsiDriver() error {
+	existingCsiDriver := &storagev1.CSIDriver{}
+	existingCsiDriver.Name = r.driver.Name
+
+	log := r.log.WithValues("driverName", existingCsiDriver.Name)
+	log.Info("Reconciling CSI Driver")
+
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(existingCsiDriver), existingCsiDriver); client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Failed to load CSI Driver resource")
+		return err
+	}
+
+	desiredCsiDriver := existingCsiDriver.DeepCopy()
```

Comment:
I'm not sure whether compiler inlines this or not, rather than sharing up, how about
``` go
desiredCsiDriver := &storagev1.CSIDriver{}
existingCsiDriver.DeepCopyInto(desiredCsiDriver)
```

Reply:
This code suffered from other issues and is completely different now. 
Resolving.